src/main/scala/xiangshan/mem/lsqueue/LoadQueueReplay.scala

```diff
@@ -339,21 +351,38 @@ class LoadQueueReplay(implicit p: Parameters) extends XSModule
   
   // generate replay mask
-  val loadReplaySelMask = VecInit((0 until LoadQueueReplaySize).map(i => {
-    val blocked = selBlocked(i) || blockByTlbMiss(i) || blockByForwardFail(i) || blockByCacheMiss(i) || blockByWaitStore(i) || blockByOthers(i)
+  val loadHigherPriorityReplaySelMask = VecInit((0 until LoadQueueReplaySize).map(i => {
+    val blocked = blockByForwardFail(i) || blockByCacheMiss(i) || blockByTlbMiss(i)
     allocated(i) && sleep(i) && !blocked && !loadCancelSelMask(i)
   })).asUInt // use uint instead vec to reduce verilog lines
-  val oldestPtr = VecInit((0 until CommitWidth).map(x => io.ldWbPtr + x.U))
-  val oldestSelMask = VecInit((0 until LoadQueueReplaySize).map(i => {
-    loadReplaySelMask(i) && VecInit(oldestPtr.map(_ === uop(i).lqIdx)).asUInt.orR
+  val loadLowerPriorityReplaySelMask = VecInit((0 until LoadQueueReplaySize).map(i => {
+    val blocked = selBlocked(i)  || blockByWaitStore(i) || blockByRARReject(i) || blockByRAWReject(i) || blockByOthers(i)
+    allocated(i) && sleep(i) && !blocked && !loadCancelSelMask(i)
   })).asUInt // use uint instead vec to reduce verilog lines
-  val remReplaySelVec = VecInit(Seq.tabulate(LoadPipelineWidth)(rem => getRemBits(loadReplaySelMask)(rem)))
-  val remOldestSelVec = VecInit(Seq.tabulate(LoadPipelineWidth)(rem => getRemBits(oldestSelMask)(rem)))
+  val loadNormalReplaySelMask = loadLowerPriorityReplaySelMask | loadHigherPriorityReplaySelMask
+  val remNormalReplaySelVec = VecInit((0 until LoadPipelineWidth).map(rem => getRemBits(loadNormalReplaySelMask)(rem)))
+  val loadPriorityReplaySelMask = Mux(loadHigherPriorityReplaySelMask.orR, loadHigherPriorityReplaySelMask, loadLowerPriorityReplaySelMask)
+  val remPriorityReplaySelVec = VecInit((0 until LoadPipelineWidth).map(rem => getRemBits(loadPriorityReplaySelMask)(rem)))
+
+  /******************************************************************************************
+   * WARNING: Make sure that OldestSelectStride must less than or equal stages of load unit.*
+   ******************************************************************************************
+   */
+  val OldestSelectStride = 4
+  val oldestPtrExt = (0 until OldestSelectStride).map(i => io.ldWbPtr + i.U)
+  val oldestMatchMaskVec = (0 until LoadQueueReplaySize).map(i => (0 until OldestSelectStride).map(j => loadNormalReplaySelMask(i) && uop(i).lqIdx === oldestPtrExt(j)))
+  val remOldsetMatchMaskVec = (0 until LoadPipelineWidth).map(rem => getRemSeq(oldestMatchMaskVec.map(_.take(1)))(rem))
+  val remOlderMatchMaskVec = (0 until LoadPipelineWidth).map(rem => getRemSeq(oldestMatchMaskVec.map(_.drop(1)))(rem))
+  val remOldestSelVec = VecInit(Seq.tabulate(LoadPipelineWidth)(rem => {
+    VecInit((0 until LoadQueueReplaySize / LoadPipelineWidth).map(i => {
+      Mux(VecInit(remOldsetMatchMaskVec(rem).map(_(0))).asUInt.orR, remOldsetMatchMaskVec(rem)(i)(0), remOlderMatchMaskVec(rem)(i).reduce(_|_))
+    })).asUInt
+  }))
```

Comment:
- The LoadQueueReplay entry will not be released immediately after the instruction is replayed from LoadQueueReplay. For example, after instruction **_a_** is replayed from LoadQueueReplay, entry **_1_** is still valid. If instruction **_a_** still needs to be replayed, entry **_1_** will be updated again, otherwise entry **_1_** can be released.
- If only the time of the first enqueue is used to select replay instructions (age matrix), when there are too many instructions (in LoadQueueReplay) to be replay, some instructions may not be selected.
- Using the pointer **_ldWbPtr_** of the oldest instruction, when the saved **_lqIdx_** of the instruction is equal to **_ldWbPtr_** and can be replayed, LoadQueueReplay will give priority to the instruction instead of using the selection result of the age matrix. In order to select more older instructions, LoadQueueReplay will calculate pointers such as **_ldWbPtr_**, **_ldWbPtr+1_**, **_ldWbPtr+2_**, **_ldWbPtr+3_**..., and if the lqIdx of the instruction is in these results, it will be selected first.
- When the pointer compared, there will be an n-bit long mask, and LoadQueueReplay will be from 0 to n-1. When **_i_** th bit is valid, select **_i_** th instruction.
- The stride of the pointer comparison is larger than the number of pipeline stages of the load unit, and the selected instruction still needs to be replayed after the first replay (for example, the data is not ready). Worse, in the bit of the mask generated by pointer comparison, the instructions (**_lqIdx_** is **_ldWbPtr+1_**, **_ldWbPtr+2_**, ...)  after the oldest instruction (**_lqIdx_** is equal to **_ldWbPtr_**)  are in the lower bit and the oldest instruction is in the higher bit. It will cannot select the oldest instruction.